core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala

```diff
@@ -1473,6 +1477,54 @@ class TimeSeriesShard(val ref: DatasetRef,
     })
   }
 
+  private def startHeadroomTask(sched: Scheduler): Unit = {
+    sched.scheduleWithFixedDelay(1, 1, TimeUnit.MINUTES, new Runnable {
+      def run() = ensureFreeBlocks
+    })
+  }
+
+  /**
+    * Expected to be called via a background task, to periodically ensure that enough blocks
+    * are free for new allocations. This helps prevent ODP activity from reclaiming immediately
+    * from itself.
+    */
+  private def ensureFreeBlocks(): Unit = {
+    val stamp = tryExclusiveReclaimLock()
+    if (stamp == 0) {
+      logger.warn(s"shard=$shardNum: ensureFreeBlocks timed out: ${reclaimLock}")
+    } else {
+      val numFree = try {
+        blockStore.ensureFreePercent(storeConfig.ensureHeadroomPercent)
+      } finally {
+        reclaimLock.unlockWrite(stamp)
+      }
+      val numBytes = numFree * blockStore.blockSizeInBytes
+      logger.debug(s"shard=$shardNum: ensureFreeBlocks: $numFree ($numBytes bytes)")
+    }
+  }
+
+  private def tryExclusiveReclaimLock(): Long = {
+    // Attempting to acquire the exclusive lock must wait for concurrent queries to finish, but
+    // waiting will also stall new queries from starting. To protect against this, attempt with
+    // a timeout to let any stalled queries through. To prevent starvation of the exclusive
```

Comment:
"let stalled queries through" Please add details. We assume that these stalled queries are liekly small ones that would not not hog the lock for longer.